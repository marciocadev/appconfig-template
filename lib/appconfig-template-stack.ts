import { CfnResource, Stack, StackProps } from 'aws-cdk-lib';
import { CfnApplication, CfnConfigurationProfile, CfnDeployment, 
  CfnDeploymentStrategy, CfnEnvironment, CfnHostedConfigurationVersion 
} from 'aws-cdk-lib/aws-appconfig';
import { Construct } from 'constructs';
import * as devConfiguration from './configurations/dev.json';
import * as hmgConfiguration from './configurations/hmg.json';
import * as prdConfiguration from './configurations/prd.json';


export class AppconfigTemplateStack extends Stack {
  application:CfnResource;
  confProfile:CfnResource;
  deploymentStrategy:CfnResource

  devEnv:CfnResource;
  hmgEnv:CfnResource;
  prdEnv:CfnResource;

  devHostedConf:CfnResource;
  hmgHostedConf:CfnResource;
  prdHostedConf:CfnResource;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    this.application = new CfnApplication(this, 'BasicApplication', {
      name: 'basic-application',
      description: 'My basic AppConfig application'
    });

    this.confProfile = new CfnConfigurationProfile(this, 'ConfigurationProfile', {
      name: 'basic-configuration-profile',
      applicationId: this.application.ref,
      locationUri: 'hosted',
      description: 'My basic AppConfig configuration profile example'
    });

    this.deploymentStrategy = new CfnDeploymentStrategy(this, 'DeploymentStrategy', {
      name: 'basic-deployment-strategy',
      deploymentDurationInMinutes: 0,
      growthFactor: 100,
      finalBakeTimeInMinutes: 0,
      replicateTo: 'NONE',
      growthType: 'LINEAR',
      description: 'My basic AppConfig deployment strategy example'
    });

    // Create environments (DEV, HMG and PRD)
    this.devEnv = new CfnEnvironment(this, 'devConfiguration', {
      name: 'dev', 
      applicationId: this.application.ref
    });
    this.hmgEnv = new CfnEnvironment(this, 'hmgConfiguration', {
      name: 'hmg',
      applicationId: this.application.ref
    });
    this.prdEnv = new CfnEnvironment(this, 'prdConfiguration', {
      name: 'prd',
      applicationId: this.application.ref
    });

    // Create hosted configuration for each environment
    this.devHostedConf = new CfnHostedConfigurationVersion(this, 'DevHostedConfig', {
      applicationId: this.application.ref,
      configurationProfileId: this.confProfile.ref,
      contentType: 'application/json',
      content: JSON.stringify(devConfiguration),
      description: 'My basic AppConfig dev hosted configuration version example'
    });
    this.hmgHostedConf = new CfnHostedConfigurationVersion(this, 'HmgHostedConfig', {
      applicationId: this.application.ref,
      configurationProfileId: this.confProfile.ref,
      contentType: 'application/json',
      content: JSON.stringify(hmgConfiguration),
      description: 'My basic AppConfig hmg hosted configuration version example'
    });
    this.prdHostedConf = new CfnHostedConfigurationVersion(this, 'PrdHostedConfig', {
      applicationId: this.application.ref,
      configurationProfileId: this.confProfile.ref,
      contentType: 'application/json',
      content: JSON.stringify(prdConfiguration),
      description: 'My basic AppConfig prd hosted configuration version example'
    });

    // Deploy environments
    new CfnDeployment(this, 'DevDeployment', {
      applicationId: this.application.ref,
      configurationProfileId: this.confProfile.ref,
      configurationVersion: this.devHostedConf.ref,
      deploymentStrategyId: this.deploymentStrategy.ref,
      environmentId: this.devEnv.ref
    });
    new CfnDeployment(this, 'HmgDeployment', {
      applicationId: this.application.ref,
      configurationProfileId: this.confProfile.ref,
      configurationVersion: this.hmgHostedConf.ref,
      deploymentStrategyId: this.deploymentStrategy.ref,
      environmentId: this.hmgEnv.ref
    });
    new CfnDeployment(this, 'PrdDeployment', {
      applicationId: this.application.ref,
      configurationProfileId: this.confProfile.ref,
      configurationVersion: this.prdHostedConf.ref,
      deploymentStrategyId: this.deploymentStrategy.ref,
      environmentId: this.prdEnv.ref
    });
  }
}
