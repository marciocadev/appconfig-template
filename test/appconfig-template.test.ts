import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { AppconfigTemplateStack } from '../lib/appconfig-template-stack';


describe('AppConfig Template Tests', () => {
    let app:App;
    let stack:AppconfigTemplateStack;
    let template:Template;

    let applicationId:string;
    let configProfileId:string;
    let deploymentStrategyId:string;

    let devHostedConfigId:string;
    let hmgHostedConfigId:string;
    let prdHostedConfigId:string;

    let devEnvId:string;
    let hmgEnvId:string;
    let prdEnvId:string;

    beforeAll(() => {
        // GIVEN
        app = new App();
        // WHEN
        stack = new AppconfigTemplateStack(app, 'MyAppConfigTemplateTest');
        // THEN
        template = Template.fromStack(stack);

        applicationId = stack.resolve(stack.application.node.id);
        configProfileId = stack.resolve(stack.confProfile.node.id);
        deploymentStrategyId = stack.resolve(stack.deploymentStrategy.node.id);

        devHostedConfigId = stack.resolve(stack.devHostedConf.node.id);
        hmgHostedConfigId = stack.resolve(stack.hmgHostedConf.node.id);
        prdHostedConfigId = stack.resolve(stack.prdHostedConf.node.id);

        devEnvId = stack.resolve(stack.devEnv.node.id);
        hmgEnvId = stack.resolve(stack.hmgEnv.node.id);
        prdEnvId = stack.resolve(stack.prdEnv.node.id);
    });

    test(('Snapshot validation'), () => {
        expect(template).toMatchSnapshot();
    });

    test(('AppConfig application created'), () => {
        template.resourceCountIs('AWS::AppConfig::Application', 1);
        template.hasResourceProperties('AWS::AppConfig::Application', {
            Name: 'basic-application'
        });
    });

    test('AppConfig Configuration Profile Created', () => {
        template.resourceCountIs('AWS::AppConfig::ConfigurationProfile', 1);
        template.hasResourceProperties('AWS::AppConfig::ConfigurationProfile', {
            Name: 'basic-configuration-profile',
            LocationUri: 'hosted',
            ApplicationId: {
                Ref: applicationId
            }
        });
    });

    test('AppConfig Deployment Strategy Created', () => {
        template.resourceCountIs('AWS::AppConfig::DeploymentStrategy', 1);
        template.hasResourceProperties('AWS::AppConfig::DeploymentStrategy', {
            Name: 'basic-deployment-strategy',
            DeploymentDurationInMinutes: 0,
            GrowthFactor: 100,
            FinalBakeTimeInMinutes: 0,
            ReplicateTo: 'NONE',
            GrowthType: 'LINEAR'
        });
    });

    test('AppConfig Environment count', () => {
        template.resourceCountIs('AWS::AppConfig::Environment', 3);
    });

    test('AppConfig DEV Environment Created', () => {
        template.hasResourceProperties('AWS::AppConfig::Environment', {
            Name: 'dev',
            ApplicationId: {
                Ref: applicationId
            }
        });
    });

    test('AppConfig HMG Environment Created', () => {
        template.hasResourceProperties('AWS::AppConfig::Environment', {
            Name: 'hmg',
            ApplicationId: {
                Ref: applicationId
            }
        });
    });

    test('AppConfig PRD Environment Created', () => {
        template.hasResourceProperties('AWS::AppConfig::Environment', {
            Name: 'prd',
            ApplicationId: {
                Ref: applicationId
            }
        });
    });

    test('AppConfig Hosted Configuration Version count', () => {
        template.resourceCountIs('AWS::AppConfig::HostedConfigurationVersion', 3);
    });

    test('AppConfig DEV Hosted Configuration Version Created', () => {
        template.findResources('AWS::AppConfig::HostedConfigurationVersion');
        template.hasResourceProperties('AWS::AppConfig::HostedConfigurationVersion', {
            ApplicationId: {
                Ref: applicationId
            },
            ConfigurationProfileId: {
                Ref: configProfileId
            },
            ContentType: 'application/json',
            Content: JSON.stringify({env: 'dev'})
        });
    });

    test('AppConfig HMG Hosted Configuration Version Created', () => {
        template.hasResourceProperties('AWS::AppConfig::HostedConfigurationVersion', {
            ApplicationId: {
                Ref: applicationId
            },
            ConfigurationProfileId: {
                Ref: configProfileId
            },
            ContentType: 'application/json',
            Content: JSON.stringify({env: 'hmg'})
        });
    });

    test('AppConfig PRD Hosted Configuration Version Created', () => {
        template.hasResourceProperties('AWS::AppConfig::HostedConfigurationVersion', {
            ApplicationId: {
                Ref: applicationId
            },
            ConfigurationProfileId: {
                Ref: configProfileId
            },
            ContentType: 'application/json',
            Content: JSON.stringify({env: 'prd'})
        });
    });

    test('AppConfig Deployment count', () => {
        template.resourceCountIs('AWS::AppConfig::Deployment', 3);
    });

    test('AppConfig DEV Deployment Created', () => {
        template.hasResourceProperties('AWS::AppConfig::Deployment', {
            ApplicationId: {
                Ref: applicationId
            },
            ConfigurationProfileId: {
                Ref: configProfileId
            },
            ConfigurationVersion: {
                Ref: devHostedConfigId
            },
            DeploymentStrategyId: {
                Ref: deploymentStrategyId
            },
            EnvironmentId: {
                Ref: devEnvId
            }
        });
    });

    test('AppConfig HMG Deployment Created', () => {
        template.hasResourceProperties('AWS::AppConfig::Deployment', {
            ApplicationId: {
                Ref: applicationId
            },
            ConfigurationProfileId: {
                Ref: configProfileId
            },
            ConfigurationVersion: {
                Ref: hmgHostedConfigId
            },
            DeploymentStrategyId: {
                Ref: deploymentStrategyId
            },
            EnvironmentId: {
                Ref: hmgEnvId
            }
        });
    });

    test('AppConfig PRD Deployment Created', () => {
        template.hasResourceProperties('AWS::AppConfig::Deployment', {
            ApplicationId: {
                Ref: applicationId
            },
            ConfigurationProfileId: {
                Ref: configProfileId
            },
            ConfigurationVersion: {
                Ref: prdHostedConfigId
            },
            DeploymentStrategyId: {
                Ref: deploymentStrategyId
            },
            EnvironmentId: {
                Ref: prdEnvId
            }
        });
    });
});
