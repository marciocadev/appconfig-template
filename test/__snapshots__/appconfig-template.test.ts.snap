// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppConfig Template Tests Snapshot validation 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "BasicApplication": Object {
      "Properties": Object {
        "Description": "My basic AppConfig application",
        "Name": "basic-application",
      },
      "Type": "AWS::AppConfig::Application",
    },
    "ConfigurationProfile": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "Description": "My basic AppConfig configuration profile example",
        "LocationUri": "hosted",
        "Name": "basic-configuration-profile",
      },
      "Type": "AWS::AppConfig::ConfigurationProfile",
    },
    "DeploymentStrategy": Object {
      "Properties": Object {
        "DeploymentDurationInMinutes": 0,
        "Description": "My basic AppConfig deployment strategy example",
        "FinalBakeTimeInMinutes": 0,
        "GrowthFactor": 100,
        "GrowthType": "LINEAR",
        "Name": "basic-deployment-strategy",
        "ReplicateTo": "NONE",
      },
      "Type": "AWS::AppConfig::DeploymentStrategy",
    },
    "DevDeployment": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "ConfigurationProfileId": Object {
          "Ref": "ConfigurationProfile",
        },
        "ConfigurationVersion": Object {
          "Ref": "DevHostedConfig",
        },
        "DeploymentStrategyId": Object {
          "Ref": "DeploymentStrategy",
        },
        "EnvironmentId": Object {
          "Ref": "devConfiguration",
        },
      },
      "Type": "AWS::AppConfig::Deployment",
    },
    "DevHostedConfig": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "ConfigurationProfileId": Object {
          "Ref": "ConfigurationProfile",
        },
        "Content": "{\\"env\\":\\"dev\\"}",
        "ContentType": "application/json",
        "Description": "My basic AppConfig dev hosted configuration version example",
      },
      "Type": "AWS::AppConfig::HostedConfigurationVersion",
    },
    "HmgDeployment": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "ConfigurationProfileId": Object {
          "Ref": "ConfigurationProfile",
        },
        "ConfigurationVersion": Object {
          "Ref": "HmgHostedConfig",
        },
        "DeploymentStrategyId": Object {
          "Ref": "DeploymentStrategy",
        },
        "EnvironmentId": Object {
          "Ref": "hmgConfiguration",
        },
      },
      "Type": "AWS::AppConfig::Deployment",
    },
    "HmgHostedConfig": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "ConfigurationProfileId": Object {
          "Ref": "ConfigurationProfile",
        },
        "Content": "{\\"env\\":\\"hmg\\"}",
        "ContentType": "application/json",
        "Description": "My basic AppConfig hmg hosted configuration version example",
      },
      "Type": "AWS::AppConfig::HostedConfigurationVersion",
    },
    "PrdDeployment": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "ConfigurationProfileId": Object {
          "Ref": "ConfigurationProfile",
        },
        "ConfigurationVersion": Object {
          "Ref": "PrdHostedConfig",
        },
        "DeploymentStrategyId": Object {
          "Ref": "DeploymentStrategy",
        },
        "EnvironmentId": Object {
          "Ref": "prdConfiguration",
        },
      },
      "Type": "AWS::AppConfig::Deployment",
    },
    "PrdHostedConfig": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "ConfigurationProfileId": Object {
          "Ref": "ConfigurationProfile",
        },
        "Content": "{\\"env\\":\\"prd\\"}",
        "ContentType": "application/json",
        "Description": "My basic AppConfig prd hosted configuration version example",
      },
      "Type": "AWS::AppConfig::HostedConfigurationVersion",
    },
    "devConfiguration": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "Name": "dev",
      },
      "Type": "AWS::AppConfig::Environment",
    },
    "hmgConfiguration": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "Name": "hmg",
      },
      "Type": "AWS::AppConfig::Environment",
    },
    "prdConfiguration": Object {
      "Properties": Object {
        "ApplicationId": Object {
          "Ref": "BasicApplication",
        },
        "Name": "prd",
      },
      "Type": "AWS::AppConfig::Environment",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
